### ISSUE TITLE
Size saving opportunity in glibc include directories

### ISSUE THREAD

**Issue by alexrp** [ziglang/zig#21258]
Right now, we have a directory for each glibc-based target triple under [`lib/libc/include`](https://github.com/ziglang/zig/tree/master/lib/libc/include). But if you do a recursive diff between many (all?) of the related ones (`riscv32-linux-gnu` vs `riscv64-linux-gnu`, `arm-linux-gnueabi` vs `arm-linux-gnueabihf`, and so on), it becomes apparent that the only difference is the presence of the appropriate `lib-names-<abi>.h` and `stubs-<abi>.h` headers. The appropriate version of this header is picked by the `lib-names.h` and `stubs.h` files based on preprocessor defines.

Also, we already have a bunch of logic for picking the right paths based on target info:

https://github.com/ziglang/zig/blob/e084c46ed6906dc51724d99b29f0992272384f5a/src/glibc.zig#L416-L663

Given these facts, I think we could enhance `process_headers.zig` to exploit this knowledge and merge these include directories together (while asserting for safety that there are no actual diffs between them). Then we'd just update `src/glibc.zig` to have very slightly smarter include directory selection.

**Comment by wooster0** [comment]
Sounds a bit like <https://github.com/ziglang/universal-headers>?

**Comment by alexrp** [comment]
Not exactly. The idea there is to have a single set of headers to cover everything. Here I'm just talking about merging headers for targets when they're literally identical already.

**Comment by alexrp** [comment]
Looks like we can delete somewhere in the ballpark of ~550 headers by doing this, which takes `lib/libc/include/*gnu*` from 2.9M to 1.3M.

**Comment by alexrp** [comment]
Note to self, these groups of directories can be merged:

* `lib/libc/include/{aarch64,aarch64_be}-linux-gnu`
* `lib/libc/include/{arm,armeb}-linux-{gnueabi,gnueabihf}`
* `lib/libc/include/csky-linux-{gnueabi,gnueabihf}`
    * Caveat: `gnu/lib-names.h` and `gnu/stubs.h` will need manual patching for float ABI.
* `lib/libc/include/loongarch64-linux-{gnu,gnusf}`
* `lib/libc/include/{{mips,mipsel}-linux-{gnueabi,gnueabihf},{mips64,mips64el}-linux-{gnuabi64,gnuabin32}}`
* `lib/libc/include/powerpc-linux-{gnueabi,gnueabihf}`
* `lib/libc/include/{powerpc-linux-{gnueabi,gnueabihf},{powerpc64,powerpc64le}-linux-gnu}`
    * Caveat: `bits/long-double.h` may need manual patching for `long double` ABI on `powerpc64le`.
* `lib/libc/include/{riscv32,riscv64}-linux-gnu`
* `lib/libc/include/{sparc,sparc64}-linux-gnu`
* `lib/libc/include/{x86-linux-gnu,x86_64-linux-{gnu,gnux32}}`

**Comment by alexrp** [comment]
> Looks like we can delete somewhere in the ballpark of ~550 headers by doing this, which takes `lib/libc/include/*gnu*` from ~2.9M to ~1.3M.

Correction: 616 headers can be removed, for a size reduction from 2.9M to 1.2M.

**Comment by alexrp** [comment]
Semi-related: `lib/libc/include/arm-netbsd-{eabi,eabihf}` and `lib/libc/include/mips-netbsd-{eabi,eabihf}` can be merged for some modest wins.
